### DISCUSSION TITLE
Referencing existing globals which may not exist + exceptions - examples

### DISCUSSION THREAD

**Discussion by btsimonh** [tasmota/Berry_playground#7] [Developer-Tools]
Thanks to @sfromis, @Jason2866 and @s-hadinger for information helping me to remove an existing driver before adding it again.

When coding in Berry, you can paste code into the Berry console, and run it.  If such code adds a driver to tasmota, for example, every time you paste, you will get an extra driver. (there will be other circumstances too where you create some global resource).

If, for example, you added your driver by using a berry variable D1, there are two ways we've found to remove the driver:

1/ use global.D1 - global variables exist on the global object, and if referenced via the global object, do not throw a comile time exception, and returns falsy if the variable does not exist at runtime.
i.e. you can detect and remove your driver using:
```
if global.D1
  tasmota.remove_driver(global.D1)
end
D1 = MyMethods()
tasmota.add_driver(D1)
```

2/ use of exceptions. - note that a simple reference to D1 will throw a COMPILE TIME exception, which you can't catch in your code.
i.e. this does not work:
```
try
  # throws 'syntax_error', but in the compiler, so code never compiles.
  print(badvar)
except .. as e,m
 print("badvar does not exist - but this wil never happen")
end
```

to get around this, you can use the compile() function.
i.e. to remove a driver using exceptions:
```
# if this is second run, remove the existing driver.
try
  var fn = compile("tasmota.remove_driver(D1)");
  print("removing existing driver")
  fn()
except .. as e,m
  print('existing driver as D1 did not exist:',e,m)
end
D1 = MyMethods()
tasmota.add_driver(D1)
```
Using exceptions to control this is not better than using global.D1, but it is a good working illustration of compile().

**Comment by s-hadinger** [comment]
Avoid `compile` like hell. This is very heavy and dangerous. Use only on last resort. This is in no way a good practice and it is highly discouraged. 

Why would you ever want to do this for a global variable?

**Reply by btsimonh** [reply]
haha...  yep, I had anticipated that compile() would not be great.  But for this use case, it would be used once at startup, or at driver replacement.  But the global.xxx construct is MUCH better.

The reasoning behind wanting to know if a global exists was, for me, this:

I was developiong a simple driver ( <20 lines ), and kept the code in a text editor.  To test, I was pasting it in the Berry console, and running it.
The driver added some web buttons. So every time I pasted, I got two additional more web buttons...  i.e. it added ANOTHER driver, the old one remaining.  I did NOT want to restart for every test.
Although tasmota.remove_driver() exists, it takes the instance reference (i.e. an EXISTING global variable), and I did not find samples.  But the First time you paste the code after boot, the global variable created as the instance reference does not exist...  hence the need to test if it does exist before trying to use it...
I first looked at exceptions - because I was getting an exception ('syntax_error' at compile).  But then later found global.xxx

br,
Simon
